tsdb/engine/tsm1/cache.go

```diff
@@ -0,0 +1,124 @@
+package tsm1
+
+import (
+	"sync"
+)
+
+// entry is the set of all values received for a given key.
+type entry struct {
+	mu         sync.Mutex
+	values     Values // All stored values.
+	checkpoint uint64 // The checkpoint associated with the latest addition to values
+	unsorted   bool   // Whether the data requires sorting and deduping before query.
+}
+
+func newEntry() *entry {
+	return &entry{values: make(Values, 0)}
+}
+
+// add adds the given values to the entry.
+func (e *entry) add(values []Value, chk uint64) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	for _, v := range values {
+		e.values = append(e.values, v)
+		// Only mark unsorted if not already marked.
+		if !e.unsorted {
+			e.unsorted = e.values[len(values)-1].Time().UnixNano() >= v.Time().UnixNano()
+		}
+	}
+	e.checkpoint = chk
+}
+
+// dedupe prepares the entry for query.
+func (e *entry) dedupe() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if !e.unsorted {
+		// Nothing to do.
+		return
+	}
+
+	return
+}
+
+// Cache maintains an in-memory store of Values for a set of keys. As data is added to the cache
+// it will evict older data as necessary to make room for the new entries.
+type Cache struct {
+	mu sync.RWMutex
+
+	checkpoint uint64
+	store      map[string]*entry
+	maxSize    uint64
+}
+
+// NewCache returns an instance of a cache which will use a maximum of maxSize bytes of memory.
+func NewCache(maxSize uint64) *Cache {
+	return &Cache{
+		store:   make(map[string]*entry),
+		maxSize: maxSize,
+	}
+}
+
+// WriteKey writes the set of values for the key to the cache. It associates the data with
+// the given checkpoint. This function is goroutine-safe and different keys can be updated
+// concurrently.
+func (c *Cache) Write(key string, values []Value, checkpoint uint64) error {
```

Comment:
The existing tsm1 code takes `points`, so this interface assumes the points have been broken apart already, and the keys formed. In other words, the cache does not expect any particular key format e.g. measurement+tagset+field-name. That policy is left to elsewhere.